convert_lora_to_gguf.py

```diff
@@ -0,0 +1,145 @@
+        "f32": gguf.LlamaFileType.ALL_F32,
+        "f16": gguf.LlamaFileType.MOSTLY_F16,
+        "bf16": gguf.LlamaFileType.MOSTLY_BF16,
+        "q8_0": gguf.LlamaFileType.MOSTLY_Q8_0,
+    }
+    ftype = ftype_map[args.outtype]
+
+    dir_base_model = args.base
+    dir_lora = args.lora_path
+    input_json = os.path.join(dir_lora, "adapter_config.json")
+    input_model = os.path.join(dir_lora, "adapter_model.bin")
+    if args.outfile is not None:
+        fname_out = args.outfile
+    else:
+        # output in the same directory as the model by default
+        fname_out = dir_lora / 'ggml-lora-{ftype}.gguf'
+
+    if os.path.exists(input_model):
+        lora_model = torch.load(input_model, map_location="cpu")
+    else:
+        input_model = os.path.join(dir_lora, "adapter_model.safetensors")
+        # lazy import load_file only if lora is in safetensors format.
+        from safetensors.torch import load_file
+        lora_model = load_file(input_model, device="cpu")
+
+    # load base model
+    logger.info(f"Loading base model: {dir_base_model.name}")
+    hparams = Model.load_hparams(dir_base_model)
+    with torch.inference_mode():
+        try:
+            model_class = Model.from_model_architecture(hparams["architectures"][0])
+        except NotImplementedError:
+            logger.error(f"Model {hparams['architectures'][0]} is not supported")
+            sys.exit(1)
+
+        model_instance = model_class(dir_base_model, ftype, fname_out, args.bigendian, False, False, None)
+        logger.info("Set model parameters")
+        model_instance.set_gguf_parameters()
+
+        # adapter_config = json.load(input_json)
+        model_instance.gguf_writer.add_string("training.type", "finetune_lora")
+
+    map_tensors: dict[str, Tensor] = {}
+    for tensor_name, tensor in lora_model.items():
+        orig_name = tensor_name.replace("base_model.model.", "")
+        orig_name = orig_name.replace(".lora_A.weight", ".weight")
+        orig_name = orig_name.replace(".lora_B.weight", ".weight")
+        is_lora_a = ".lora_A.weight" in tensor_name
+        is_lora_b = ".lora_B.weight" in tensor_name
+        if not is_lora_a and not is_lora_b:
+            logger.error(f"Unexpected name '{tensor_name}': Not a lora_A or lora_B tensor")
+            sys.exit(1)
+        dest_name = model_instance.map_tensor_name(orig_name)
+        dest_name = f"{dest_name}.lora_a" if is_lora_a else f"{dest_name}.lora_b"
+        # logger.info(f"{orig_name} --> {dest_name}")
+        map_tensors[dest_name] = tensor
+
+    # overwrite method
+    def get_tensors(self) -> Iterator[tuple[str, Tensor]]:
+        for name, tensor in map_tensors.items():
+            yield (name, tensor)
+
+    # overwrite method
+    def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iterable[tuple[str, Tensor]]:
+        del bid  # unused
+        # TODO: This will not take into account tensor transformations
+        return [(name, data_torch)]
```

Comment:
This is quite a big TODO. It was the reason the previous script was removed. I think this needs to be handled on a model by model basis, and reject models not explicitly supported, otherwise it will lead to the creation of incorrect gguf lora files.

Reply:
`modify_tensors` does different things in different cases. Sometimes it filters out tensors, sometimes it splits them, sometimes it duplicates them, sometimes it permutes them, and sometimes it changes the values (like for Gemma's norms).

If lora A and B weight can be turned into a tensor with the same shape as the weight tensor they're affecting, then transformed with `modify_tensors`, then split back into A and B, then it would allow all supported models to work with this.

But I'm not sure if this is doable or not, or if the `modify_tensors` functions should instead have a lora mode or something like that.

Alternatively, there *might* be a way to make this work with some kind of `LazyLoraTensor` (although it might not be possible to use `gguf.LazyBase` as a base class because it assumes a more direct mapping of operations) by giving a fake tensor with the weight shape to the original `modify_tensors` of the `model_instance`, then doing some magic interceptions to gracefully handle shape and value transformations, to then output back the lora weights along with their names.

Reply:
> If lora A and B weight can be turned into a tensor with the same shape as the weight tensor they're affecting, then transformed with the original `modify_tensors` of the `model_instance`, then split back into A and B, then it would allow all supported models to work with this.

I thought about this, but it still be very tricky because not all transformations are the same. For example, MOE tensors are stacked into 3d tensor which totally change the shape of the output tensor.

On the other hands, simple matrix add or scale can be applied to only lora_B tensor (why keeping lora_A untouched). For row permutation, I tried with simple numpy script and observed that `LlamaModel.permute(B)*A == LlamaModel.permute(B*A)`, but not sure if it's a know mathematics property or not.

> I think this needs to be handled on a model by model basis, and reject models not explicitly supported, otherwise it will lead to the creation of incorrect gguf lora files.

Until now I fully understand why the old python script was removed. You're right, I think for now we should explicitly check if given model support lore --> gguf conversion or not. Probably adding `Model.support_lora()`, and for now enable only llama-only (without MOE, so no mixtral)

Reply:
So I modified the script to re-use `modify_tensors()` from the original model class. Also added `support_lora()`, so model not overriding this method will not be able to convert lora.

Commit: https://github.com/ggerganov/llama.cpp/pull/8332/commits/ee2b35c65f7e4e862990d6460c4b0a0ac433a874

Reply:
> sometimes it stacks them, sometimes it duplicates them, sometimes it permutes them, and sometimes it changes the values (like for Gemma's norms).

Btw, I'm trying to list out some of the transformation in `modify_tensors` that are easy to support. I'll have a look in the future:
- We don't care if transformation is used on non-lora tensors like `norm.weight` ==> gemma should be ok to support
- `LlamaModel.permute(B)*A == LlamaModel.permute(B*A)` at least observed on my tests (or maybe just coincident)
- Scale `BA*scale == (B*scale)*A`

Reply:
> then doing some magic interceptions to gracefully handle shape and value transformations

@ngxson I attempted to implement my "magic interceptions" suggestion in <https://github.com/compilade/llama.cpp/commit/8db62494f43d86908c448001c66ae375ba3d009d>.

I get exactly the same `sha256sum` when converting a Llama-3 LoRA with this and with the previous approach.

Mixtral LoRAs can now also be converted (tested conversion for <https://huggingface.co/maxkretchmer/gc-mixtral/tree/main>, the first one I found which had LoRA of the MoE and not only Attention), but I can't test inference with that on my hardware (not enough RAM for Mixtral).

I'm looking to find LoRAs of smaller models which use qkv splitting in `modify_tensors` to test this for correctness.

> `LlamaModel.permute(B)*A == LlamaModel.permute(B*A)` at least observed on my tests (or maybe just coincident)

I can confirm this does seem to be equivalent. I think it's because the reshapes in `LlamaModel.permute` never change the row size for 2D tensors, so A is left intact.

Reply:
Second thought, it would make more sense for row premutation to be `premute(B*A) == premute(B) * premute(A^T)^T`. Though, this assertion failed on my test (surprise!).

In any case, we can keep `permute(B)*A == permute(B*A)` as we have today because it's already working. We can re-visit this later if needed.

Reply:
Loras that modify the experts in MoE models will not work, that would require adding lora support in the `ggml_mul_mat_id` operation. I don't think this could be implemented just by changing the graph either, the expert used is only known during the evaluation, so this would need to be implemented in the backends. Until that is done we should avoid allowing loras that modify the experts, either by rejecting them during conversion or by adding checks during loading in llama.cpp.

Reply:
Yeah right, so I think on top of the script by @compilade , we can bring back my `Model.support_lora()` to explicitly disable support for MOE models.

Reply:
> Loras that modify the experts in MoE models will not work, that would require adding lora support in the `ggml_mul_mat_id` operation. I don't think this could be implemented just by changing the graph either, the expert used is only known during the evaluation

Hmm, yes that is a problem. But aren't the I32 ids of the chosen experts in their own tensor? I think it might be possible to make this work with `ggml_reshape` and `ggml_get_rows` on the LoRA weights in `llm_build_moe_ffn`.

> we can bring back my `Model.support_lora()` to explicitly disable support for MOE models.

There are still Attention-only LoRAs of MoE models which exist, and I don't see why these shouldn't be supported.

Reply:
That may work, but the overhead with all the copies would be pretty bad. I think that runtime loras need to be very low overhead, otherwise people will just use pre-merged models.

Reply:
> Hmm, yes that is a problem. But aren't the I32 ids of the chosen experts in their own tensor?

I'm not very familiar with `llm_build_moe_ffn`, so I drafted a naive idea like this:

```cpp
tensor up_exps_a; // [x, y, n_experts] ==> all lora_a of all experts stacked into 3D
tensor up_exps_b; // [y, x, n_experts] ==> all lora_b of all experts stacked into 3D
// lora A*cur
tensor up_a = ggml_mul_mat_id(ctx, up_exps_a, cur, selected_experts); // [?, n_expert_used, n_tokens]
// lora B*(A*cur)
tensor up_b_used = ggml_get_rows(..., up_exps_b); // [y, x, n_expert_used] ==> select only "used" experts from up_exps_b
tensor up_ab = ggml_mul_mat(ctx, up_exps_b, up_a); // [n_ff, n_expert_used, n_tokens]
```

I don't know if `ggml_get_rows` can work that way, but I agree which what @slaren is that here `ggml_get_rows` works with up_exps_b, which is quite a lot of data to copy.

(Also I have no idea how to make the dimension of `up_b_used` to be correct. Probably my draft is simply wrong)

Reply:
Actually, maybe `ggml_mul_mat_id` be directly used instead of `ggml_get_rows` and `ggml_mul_mat` for `up_exps_b`?

```c
tensor up_exps_a; // [n_embd, rank, n_experts] ==> all lora_a of all experts stacked into 3D
tensor up_exps_b; // [rank, n_embd, n_experts] ==> all lora_b of all experts stacked into 3D
// lora A*cur
tensor up_a = ggml_mul_mat_id(ctx, up_exps_a, cur, selected_experts); // [rank, n_expert_used, n_tokens]
// lora B*(A*cur)
tensor up_ab = ggml_mul_mat_id(ctx, up_exps_b, up_a, selected_experts); // [n_embd, n_expert_used, n_tokens]
```

Not yet tested, but it *seems* like it could work.

(currently downloading a [4x1b MoE](https://huggingface.co/Lambent/cosmoem-4x1b) + [LoRA](https://huggingface.co/Lambent/CosMoE-AlpacaLight-v0.6), to possibly test this on my hardware)

Reply:
That should work.

Reply:
@compilade FYI, I made a stupidly simple script to generate a MOE model with random weight.

To use it, set temperature to 0: `./llama-cli -m ./fake_moe.gguf -n 20 --temp 0`  
It should always output: `!YYYYYYYYYYYYYYYYYYYY`

What I want to do now is to generate random lora and merge it to the original model, so we can test the output of `llama-cli` with base+lora and with merged model

Reply:
Ok so I for testing that, I can finally generate a very tiny MOE model based on stories15M, generate a random lora, then merge it to the MOE. The script can be found here: https://gist.github.com/ngxson/bca68453aa411c058e6c84b7f7925223

I'm now looking into the code to see how I can load lora experts into 3D tensors

Reply:
With the script above, I generated the model+lora and implement `llm_build_lora_mm_id` in https://github.com/ggerganov/llama.cpp/pull/8332/commits/916e95928b0757fb9e5e601ee5f325af29e5253e

Surprisingly it works with the "toy" model generated from stories15M. I couldn't test with `llama-perplexity` as @compilade suggested, but would be nice to do so. (I need to sleep now)

Reply:
So today I made the huggingface version of the toy model above. It's [stories15M_MOE](https://huggingface.co/ngxson/stories15M_MOE)

On top of that, I also trained a [shakespeare15M_MOE LoRA version](https://huggingface.co/ngxson/stories15M_MOE/tree/main/moe_shakespeare15M/checkpoint-500) using some shakespeare texts. This can be used to test the conversion script. The "random weight" adapter I made earlier is quite noisy and hard to use. 

@compilade I didn't test on your version of the converting script, but could you give it a try? For now I don't have any method to convert that shakespeare lora to gguf.

Reply:
@ngxson Thanks a lot for this! I've converted the LoRA with my convert script, and I'm getting the same text as the sample output, but only when I scale the LoRA by 2, not sure if that's expected:

```console
$ ./build/bin/llama-cli --log-disable -m /srv/LLMstash/tmp/stories_MOE-f16.gguf --temp 0 -p "Look in thy glass" -n 32
 Look in thy glass was a little girl. She was only three years old and she was three years old. She was three years old. She was three years old and she was
$ ./build/bin/llama-cli --log-disable -m /srv/LLMstash/tmp/stories_MOE-f16.gguf --temp 0 -p "Look in thy glass" -n 32 --lora /srv/LLMstash/tmp/lora_shakespeare_stories_MOE-f16.gguf
 Look in thy glass is a precious thing. They are so precious and can be held. They are made of glass, and they are not toys. They are made
$ ./build/bin/llama-cli --log-disable -m /srv/LLMstash/tmp/stories_MOE-f16.gguf --temp 0 -p "Look in thy glass" -n 128 --lora-scaled /srv/LLMstash/tmp/lora_shakespeare_stories_MOE-f16.gguf 2
 Look in thy glass in love of the eye:
That's when when the eye see thy on the sun's shining bright.
That's why those flowers look so bright,
But then thy in our glass's line,
Whose eye is red, then feasting on the other half,
When in their crows are red,
Whose eye is red, then on the so eye doth a side,
O in touching red thy a red eye doth a different,
And for my self I'll be a flower's sun's sun,
And for a red side
```

Reply:
@compilade great, thanks! The result looks correct.

> I scale the LoRA by 2

I fine tuned with rank 64 and alpha 128 so `scale = 128/64 = 2` is correct.

In fact, I should have taken the rank and alpha from `adapter_config.json` then save it to gguf, so that when load we can automatically set the scale. I did not do that because in theory, we could have different rank for each tensors. But after second thought, in practical, most fine tuning frameworks use the same rank for all tensors, so we should not have any problems with my initial idea.

Reply:
@ngxson I've finally pushed to your branch the modifications making `convert_lora_to_gguf.py` more general.

> In fact, I should have taken the rank and alpha from `adapter_config.json` then save it to gguf, so that when load we can automatically set the scale.

I've added the `alpha` as `training.lora.alpha` to the GGUFs made by `convert_lora_to_gguf.py` in <https://github.com/ggerganov/llama.cpp/pull/8332/commits/8956543c091c6851089ed7467fa44ac2b2b0ee37>. I did not store the rank (yet) because it can be taken from the tensor dimensions, but this can still be changed relatively easily.

I've also made pretty much every model graph use `llm_build_lora_mm` instead of `ggml_mul_mat` where it made sense in <https://github.com/ggerganov/llama.cpp/pull/8332/commits/87301bdd59554604ce0103fe39580a1608cf97cd>. Otherwise LoRA adapters for other architectures than `llama` were not really applied.

Reply:
Perfect, thanks @compilade !

> I did not store the rank (yet) because it can be taken from the tensor dimensions, but this can still be changed relatively easily.

After thinking more about it, I think we can add a kv `training.lora.scale` to store the default scale value. For now it will be calculated by alpha/rank, and in the future if users use another training framework, they can add their own logic to calculate default scale (so should be future-proof)

Reply:
Oh sorry I didn't see your modification inside `llm_build_lora_mm`. Yeah calculating scale differently for each tensor is better. I'm reverting my change https://github.com/ggerganov/llama.cpp/pull/8332/commits/42415a4874e0f963e4aca6796ea5dfb97cd17464